src/cdk/scrolling/virtual-for-of.ts

```diff
@@ -0,0 +1,310 @@
+   * @param index The index of the data element whose client rect we want to measure.
+   * @return The combined client rect for all DOM elements rendered as part of the given index.
+   *     Or null if no DOM elements are rendered for the given index.
+   * @throws If the given index is not in the rendered range.
+   */
+  measureClientRect(index: number): ClientRect | null {
+    if (index < this._renderedRange.start || index >= this._renderedRange.end) {
+      throw Error('Error: attempted to measure an element that isn\'t rendered.');
+    }
+    index -= this._renderedRange.start;
+    let view = this._viewContainerRef.get(index) as
+        EmbeddedViewRef<CdkVirtualForOfContext<T>> | null;
+    if (view && view.rootNodes.length) {
+      let minTop = Infinity;
+      let minLeft = Infinity;
+      let maxBottom = -Infinity;
+      let maxRight = -Infinity;
+
+      // There may be multiple root DOM elements for a single data element, so we merge their rects.
+      for (let i = 0, ilen = view.rootNodes.length; i < ilen; i++) {
+        let rect = (view.rootNodes[i] as Element).getBoundingClientRect();
+        minTop = Math.min(minTop, rect.top);
+        minLeft = Math.min(minLeft, rect.left);
+        maxBottom = Math.max(maxBottom, rect.bottom);
+        maxRight = Math.max(maxRight, rect.right);
+      }
+
+      return {
+        top: minTop,
+        left: minLeft,
+        bottom: maxBottom,
+        right: maxRight,
+        height: maxBottom - minTop,
+        width: maxRight - minLeft
+      };
+    }
+    return null;
+  }
+
+  ngDoCheck() {
+    if (this._differ && this._needsUpdate) {
+      const changes = this._differ.diff(this._renderedItems);
+      if (!changes) {
+        this._updateContext();
+      } else {
+        this._applyChanges(changes);
+      }
+      this._needsUpdate = false;
+    }
+  }
+
+  ngOnDestroy() {
+    this._viewport.disconnect();
+
+    this._dataSourceChanges.complete();
+    this.viewChange.complete();
+
+    for (let view of this._templateCache) {
+      view.destroy();
+    }
+  }
+
+  /** React to scroll state changes in the viewport. */
+  private _onRenderedRangeChange(renderedRange: Range) {
+    this._renderedRange = renderedRange;
+    this.viewChange.next(this._renderedRange);
+    this._renderedItems = this._data.slice(this._renderedRange.start, this._renderedRange.end);
+    if (!this._differ) {
+      this._differ = this._differs.find(this._renderedItems).create(this.cdkVirtualForTrackBy);
+    }
+    this._needsUpdate = true;
+  }
+
+  /** Swap out one `DataSource` for another. */
+  private _changeDataSource(oldDs: DataSource<T> | null, newDs: DataSource<T>): Observable<T[]> {
+    if (oldDs) {
+      oldDs.disconnect(this);
+    }
+    this._needsUpdate = true;
+    return newDs.connect(this);
+  }
+
+  /** Update the `CdkVirtualForOfContext` for all views. */
+  private _updateContext() {
+    for (let i = 0, len = this._viewContainerRef.length; i < len; i++) {
```

Comment:
Given the sensitivity of perf here, you might want to consider a faster loop technique. https://jsperf.com/loops/33